dependencies/src/main/java/com/google/cloud/tools/opensource/classpath/FixedSizeClassPathRepository.java

```diff
@@ -0,0 +1,116 @@
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package com.google.cloud.tools.opensource.classpath;
+
+import com.google.common.annotations.VisibleForTesting;
+import com.google.common.cache.Cache;
+import com.google.common.cache.CacheBuilder;
+import java.util.HashMap;
+import java.util.Map;
+import javax.annotation.Nullable;
+import org.apache.bcel.classfile.JavaClass;
+import org.apache.bcel.util.ClassPath;
+import org.apache.bcel.util.ClassPathRepository;
+
+/**
+ * This repository behaves the same as {@link ClassPathRepository} except that this class limits the
+ * size of its {@link JavaClass} cache at most {@code maximumSize} entries. When the cache reaches
+ * the limit, it evicts entries that have not been used recently.
+ *
+ * <p>This class avoids {@code OutOfMemoryError: gc overhead limit exceeded} that occurs when
+ * parsing too many JAR files to handle with {@link ClassPathRepository} or {@link
+ * org.apache.bcel.util.MemorySensitiveClassPathRepository}, while providing reasonable speed by
+ * caching frequently-used instances.
+ *
+ * <p>The default maximum size is 1000 entries. As per experiments with spring-cloud-gcp project,
+ * this maximum size gives the best performance when running {@link
+ * LinkageChecker#findSymbolProblems()}.
+ *
+ * @see <a href="https://github.com/google/guava/wiki/CachesExplained#size-based-eviction">Guava
+ *     CachesExplained: Size-based Eviction</a>
+ * @see <a href="https://github.com/GoogleCloudPlatform/cloud-opensource-java/issues/653"
+ *     >Performance evaluation of FixedSizeClassPathRepository with different maximum cache size</a>
+ */
+final class FixedSizeClassPathRepository extends ClassPathRepository {
+
+  private final Cache<String, JavaClass> loadedClass;
+
+  /**
+   * Mapping from class names to special class file locations.
+   *
+   * <p>While class name and class file name are the same in most cases, sometimes classes are not
+   * placed in the root of a JAR file to support a framework-specific JAR structure. For example,
+   * Spring Boot Gradle Java plugin places class files under "BOOT-INF/classes". To load such
+   * classes by class name, this mapping keeps track of the special location once they are loaded.
+   *
+   * <ul>
+   *   <li>Key: class name (value from {@link JavaClass#getClassName()}) which has a special class
+   *       file name different from its class name. Example: {@code com.google.Foo}
+   *   <li>Value: the special class file name as in {@link JavaClass#getFileName()}, a path that
+   *       locates a class file in a class path. Example: {@code BOOT-INF.classes.com.google.Foo}
+   * </ul>
+   *
+   * @see <a
+   *     href="https://docs.spring.io/spring-boot/docs/current/reference/html/executable-jar.html"
+   *     >Spring Boot Reference Guide: The Executable Jar Format</a>
+   */
+  private final Map<String, String> specialClassFileName;

```

Comment:
I think simply `classFileNames`. It should be plural and `special` doesn't say much now that we have the comments to describe this.